lib/server.dart

```diff
@@ -0,0 +1,142 @@
+  ///
+  /// Name of the method being called
+  /// Parameters that were passed in
+  /// Exception or error that occurred
+  final MethodFailureSignature methodFailureHandler;
+
+  /// The error code for a missing method
+  final int noExistingMethodCode;
+
+  /// The error code for malformed xml
+  final int malformedXmlCode;
+
+  /// The error code for unsupported encoding
+  final int unsupportedEncoding;
+
+  /// Creates a [XmlRpcHandler] that handles the set of [methods]
+  ///
+  /// It uses the specified set of [codecs] for encoding and decoding
+  /// The [methodErrorHandler] is a callback that handles any exception or error thrown from the methods
+  /// The code [noExistingMethodCode] is the fault code returned if the method is not found
+  XmlRpcHandler({
+    @required this.methods,
+    List<Codec> codecs,
+    MethodFailureSignature methodFailureHandler,
+    int noExistingMethodCode,
+    int malformedXmlCode,
+    int unsupportedEncoding,
+  })  : codecs = codecs ?? standardCodecs,
+        methodFailureHandler =
+            methodFailureHandler ?? _defaultMethodFailureHandler,
+        noExistingMethodCode = noExistingMethodCode ?? -32601,
+        malformedXmlCode = malformedXmlCode ?? -32700,
+        unsupportedEncoding = unsupportedEncoding ?? -32701;
+
+  /// A default method error handling callback
+  ///
+  /// The default fault code specified here is from the spec [here](http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php)
+  static Fault _defaultMethodFailureHandler(
+          String method, List<dynamic> params, dynamic error) =>
+      Fault(-32500,
+          'Server error. Calling $method, with params $params failed with error $error');
+
+  /// Marshalls the [data] from XML to Dart types, and then dispatches the function, and marshals the return value back into the XMLRPC format
+  Future<XmlDocument> handle(XmlDocument document) async {
+    var returnValue;
+    try {
+      final methodCall = document.findElements('methodCall').first;
+      final methodName = methodCall.findElements('methodName').first.text;
+      if (methods.containsKey(methodName)) {
+        final params = methodCall.findElements('params');
+        final parsedArgs = params.isNotEmpty
+            ? params.first
+                .findElements('param')
+                .map((arg) => arg.findElements('value').first)
+                .map(getValueContent)
+                .map((e) => decode(e, codecs))
+                .toList()
+            : [];
+
+        returnValue = await _dispatch(methodName, parsedArgs);
+      } else {
+        returnValue = Fault(noExistingMethodCode,
+            'Server error. No method by the name $methodName, registered with this server');
+      }
+    } on ArgumentError catch (e) {
+      returnValue = Fault(unsupportedEncoding,
+          'Parse error. Decoding arguments failed, got: ${document.toXmlString()}, error was $e');
```

Comment:
`unsupportedEncoding` is more related to char encoding (utf8...) imho.

The more I think about how to deal with those error cases the more I tend to have custom exceptions thrown and let the implementation handle them. Something like:

```dart
class XmlRpcException extends Exception {}
class XmlRpcRequestFormatException extends XmlRpcException {}
/// When an exception occurs in call
class XmlRpcCallException extends XmlRpcException {}
/// When an exception occurs in response encoding
class XmlRpcResponseEncodingException extends XmlRpcException {}
```

We could also add an helper method on `XmlRpcHandler` to encode `Fault`. This will allow to get the xml to be sent to the client.

Thus in `simple_server` we would have:

```dart
        try {
          response = (await handler.handle(parse(result))).toXmlString();
        } on XmlRpcRequestFormatException {
          response = handler.encodeFault(Fault(-32700, 'parse error. not well formed')).toXmlString();
        } on XmlRpcCallException {
          response = handler.encodeFault(Fault(-32500, 'application error')).toXmlString();
        }
```

WDYT?

Reply:
Yeah, I think that is a better approach.

If you have the time, go ahead and merge it in and make the changes, since I think you have a better idea on how you would like to expose this to the end user.

If you are busy, I'd be willing to make the adjustments from your comments.

Thanks!